sql/core/src/test/scala/org/apache/spark/sql/execution/command/DDLSuite.scala

```diff
@@ -3030,6 +3030,47 @@ abstract class DDLSuite extends QueryTest with SQLTestUtils {
       }
     }
   }
+
+  test("REFRESH FUNCTION") {
+    val msg = intercept[AnalysisException] {
+      sql("REFRESH FUNCTION md5")
+    }.getMessage
+    assert(msg.contains("Cannot refresh builtin function"))
+
+    withUserDefinedFunction("func1" -> true) {
+      sql("CREATE TEMPORARY FUNCTION func1 AS 'test.org.apache.spark.sql.MyDoubleAvg'")
+      val msg = intercept[AnalysisException] {
+        sql("REFRESH FUNCTION func1")
+      }.getMessage
+      assert(msg.contains("Cannot refresh temporary function"))
+    }
+
+    withUserDefinedFunction("func1" -> false) {
+      intercept[NoSuchFunctionException] {
+        sql("REFRESH FUNCTION func1")
+      }
+
+      val func = FunctionIdentifier("func1", Some("default"))
+      sql("CREATE FUNCTION func1 AS 'test.org.apache.spark.sql.MyDoubleAvg'")
+      assert(!spark.sessionState.catalog.isRegisteredFunction(func))
+      sql("REFRESH FUNCTION func1")
+      assert(spark.sessionState.catalog.isRegisteredFunction(func))
+
+      spark.sessionState.catalog.externalCatalog.dropFunction("default", "func1")
+      assert(spark.sessionState.catalog.isRegisteredFunction(func))
+      sql("REFRESH FUNCTION func1")
+      assert(!spark.sessionState.catalog.isRegisteredFunction(func))
+
+      val function = CatalogFunction(func, "test.non.exists.udf", Seq.empty)
+      spark.sessionState.catalog.createFunction(function, false)
```

Comment:
It's a bit weird if we don't fail invalid functions when creating, but fail when refreshing it. How hard is it to make `REFRESH TABLE` lazy?

Reply:
I don't know why we not check function during create. It seems no use to create a not exists function but can produce some problem like typo.

The same command, Hive failed directly `create function f1 as 'test.non.exists.udf'`.

Reply:
We can fix CREATE FUNCTION later and update this test.

Reply:
or is it because you are calling the internal API not the `CREATE FUNCTION` command?

Reply:
If we make `REFRESH FUNCTION` lazy as `CREATE FUNCTION`, something like this
```
if (catalog.isRegisteredFunction(identifier)) {
  catalog.unregisterFunction(identifier)
}
if (!catalog.isPersistentFunction(identifier)) {
  throw new NoSuchFunctionException(identifier.database.get, functionName)
}
```

The different thing is we don't register/check function and the register/check action happened when user query with this function like `select func(f)`.

I think it might be better to do the function check right now.

Reply:
Maybe we should make `CREATE FUNCTION` not lazy, in a new PR.

Reply:
ok, I will try it after this pr finished.